Updates to cell library generation

ttown523 opened the pull request:
## In this commit:

This commit contains several modifications to the cell library generation in TINCR. 
- A large portion of the code was refactored to be more readable. This includes using more descriptive variable and function names, splitting dedicated workloads to their own function, and updating some function implementations.
- Doxygen comments were added to each function in genCellLibrary.tcl.
- Support for configurable pin mappings was added. Currently this is not being used because it is not supported in RapidSmith, but it will be used in the future.  
- Assertions were added throughout the code to document assumptions and help debug. 
- Other updates to the cell library generation such as adding cell pin types and other cell tags.

ttown523 committed:
Updates to cell library generation

ttown523 committed:
Merge branch 'master' of https://github.com/byuccl/tincr

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -6,43 +6,54 @@ package require tincr.cad.util 0.0
     namespace export \
-        create_xml_cell_library
+        create_xml_cell_library \
+        test_cell_library
```
Tincr has a unit testing framework (see [tincr_test](https://github.com/byuccl/tincr/tree/master/tincr_test)) that should be used instead of writing dedicated test methods in production code.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -6,43 +6,54 @@ package require tincr.cad.util 0.0
+## Splits the <code>string</code> by <code>token</code>, and returns the last element in the list.
+#  Helper function used to get the relative name of Vivado elements. For example, the call
+#  <code>suffix "I/am/a/test" "/"</code> will return the string "test."
+#  TODO: add to tincr.util
```
Why not just add it to tincr.cad.util now?

ttown523 replied:
Done.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -6,43 +6,54 @@ package require tincr.cad.util 0.0
-            return 0
+# @param cell The <code>cell<code> instance to be placed
+# @param bel The <code>bel</code> to try placing <code>cell<code> on.
+proc is_my_placement_legal { cell bel } {
```
The [`cells` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/design/cells.tcl) already has a `is_placement_legal` method. Delete this method and use/update that one instead.

ttown523 replied:
Done. I updated the method in tincr::cells

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -6,43 +6,54 @@ package require tincr.cad.util 0.0
+#   TODO: find a way to merge this and get_supported_lib_cells 
+#
+# @return A list of supported leaf cells.
 proc get_supported_leaf_libcells { } {
```
This method belongs in the [`lib_cells` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/design/lib_cells.tcl).

ttown523 replied:
Done.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -62,8 +73,12 @@ proc get_supported_leaf_libcells { } {
+#   TODO: update this 
+#
+# @return A list of supported cells.
 proc get_supported_libcells { } {
```
This method belongs in the [`lib_cells` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/design/lib_cells.tcl).

ttown523 replied:
Done.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+#   to each primitive site type in the current device. This function chooses 
+#   default site locations over alternate site locations, and chooses a different
+#   site location for alternate types if possible. 
+#   TODO: update the tincr::sites::unique function
```
Why not just update the `unique` method now?

ttown523 replied:
I have updated the unique method. See the sites ensemble.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+#   TODO: update the tincr::sites::unique function
+#
+# @returns A list with two elements. <br>
+#           (1) The first element is the map from a site type to a site location <br>
```
Pretty sure we're already generating and caching this information. Check out the `array.site_type.sites` cache in [caches.tcl](https://github.com/byuccl/tincr/blob/master/tincr/cad/cache/cache_definitions/caches.tcl).

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+# @param site_map Dictionary mapping site types, to site locations
+# @param alternate_site_set Set of sites that are only alternate sites
+# @return A dictionary that maps a cell to all sites it can be placed on
+proc create_cell_to_site_map {lib_cells site_map alternate_site_set} {
```
This method belongs in the [`sites` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/device/sites.tcl). In fact, the `compatible_with` method has the exact same purpose.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
-    set should_reset 0
+# @param lib_cells List of supported library cells for the current device
+# @param site_map Dictionary mapping site types, to site locations
+# @param alternate_site_set Set of sites that are only alternate sites
```
You're passing in all these lists, dictionaries, and sets of information that really should be cached using Tincr's caching mechanism. In fact, `site_map` is already cached as `array.site_type.sites`. Using the caches will provide you with a significant performance gain.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
-                dict lappend cell_site_map $lbc $sitename
-                unplace_cell $c
+            
+            if {[catch {[place_cell $cell_instance $site]} err] == 0} {
```
You can get this same information _much_ faster using the `array.lib_cell.bel_types` cache.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+                get_static_leaf_cell_pin_mappings $cell $bel $xml_out
+            }
+            
+            if {0} { ; # START COMMENT
```
What's the point of this line? It's always false, meaning the `else` is always executed. Just remove the `if` ... `else`.

ttown523 replied:
This line is for configurable cell pin mappings. Currently RapidSmith does not support this, which is why this part of the code is commented out for now. We will need it soon, so I don't want to delete the code. I have added a comment describing why this if statement is there.

bradselw replied:
Sounds good

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+# @param site Site location to place the cell on
+# @param xml_out Output file to print the XML to
+#
+proc process_leaf_cell {cell site xml_out} {
```
Since you are writing out a file that describes design elements (cells), this method belongs somewhere in the tincr.io.design package IMO.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+# @param cell Cell instance
+# @param bel Bel to place the cell on
+# @param xml_out File handle to write the pin mapping XML
+proc get_static_leaf_cell_pin_mappings { cell bel xml_out } {
```
IMO this method belongs somewhere in the `tincr.io.design` package.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
+#
+# @returns A list with two elements. <br>
+#           (1) The first element is the map from a site type to a site location <br>
+#           (2) A set of site types that are only alternate site types <br>
```
This "set of site types that are only alternate site types" is static (it's the same every time it is generated for a given part). Each time you regenerate it is wasted effort. You should store this using Tincr's caching framework, that way you only have to generate it once because it is persisted to disk between sessions

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
-            # Since a carry4 cell is not a LUT,INV, or BUF, we cannot use the LOCK_PINS property to try and map the cell
-            # pins onto bel pins
+## Gets the default value of a configuration for a given cell
+#   TODO: add this to tincr::cell?
```
Add this to the [`cells` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/design/cells.tcl).

ttown523 replied:
Done.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -74,176 +89,583 @@ proc get_supported_libcells { } {
-            foreach cell_pin $unconnectedCellPins {
-                connect_net -net [get_nets tmp_net] -objects [get_pins $cell_pin]
+## Resets the cell to its default configuration. 
+#   TODO: add this to tincr::cell?
```
Add this to the [`cells` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/design/cells.tcl).

ttown523 replied:
Done.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+#
+# @param lib_cell Library cell
+# @param xml_out File handle to the output XML file
+proc write_property_xml { lib_cell xml_out } {
```
This belongs in `tincr.io.design`.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+#
+# @param site_map Dictionary mapping site type to a physical site location
+# @param xml_out File handle to the output XML file
+proc create_port_xml { site_map xml_out } {
```
Belongs in the `tincr.io.design` package.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
-	puts $fo "      </libcellproperties>"
+## Gets the type of the specified pin according to Vivado. DATA is the
+#   default pin type where no other pin type is specified.
+#   TODO: on each new release of Vivado, verify this function is still correct.
```
You could write a unit test in `tincr_test` to verify that the function is still correct.

ttown523 replied:
Unfortunately, the correctness of this function has to be manually verified. This is because for a cell pin, there are several pin types (as shown in the function implementation) which are based on the naming "IS_*" (for example IS_CLOCK). However, there are also other properties such as IS_CONNECTED, IS_LEAF, and IS_ORIG_PIN that are not pin types but match the pattern. There is no way to reliable filter these out in a test. I added a not in the comments saying that it has to be manually verified.

bradselw replied:
Good ol' Vivado

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+#
+# @param pin Cell pin
+# @return The type of cell pin
+proc get_pin_type { pin } {
```
Move this to the [`pins` ensemble](https://github.com/byuccl/tincr/blob/master/tincr/cad/design/pins.tcl).

ttown523 replied:
Moved.

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+# @param part Xilinx FPGA part to generate a cell library for
+# @param threshold Threshold of configurable pin mappings to compute before quitting. (currently not used
+#           but will be used in the future)
+proc ::tincr::create_xml_cell_library { {part xc7a100t-csg324-3} {filename ""} {threshold 100}} {
```
How about putting all these XML generation methods in a file `rapidsmith.tcl` inside of the `tincr.io.design` package.

ttown523 replied:
That is a good idea, but I will wait to refactor this in a later pull request.

bradselw replied:
Alright

bradselw commented on tincr/io/library/genCellLibrary.tcl:
```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+# and debugging enabled. 
+#
+# @param part Xilinx FPGA part
+proc ::tincr::test_cell_library { {part xc7a100t-csg324-3} } {
```
Make this a unit test?

bradselw reviewed (changes requested):
That was a lot to go through! My suggestions boil down to this:
- Utilize Tincr's caching mechanism for all the static data you are generating. Some of the information that you are generating already exists in a cache somewhere in Tincr. These caches provide significant performance gains because they only have to be generated once and are persisted on disk.
- Many of these new methods should be moved into other packages/ensembles to provide better encapsulation. I think you were already planning to do this, because I saw several TODOs that said as much. I would suggest refactoring them now rather than later.
- Reuse code whenever possible. A few of the methods you wrote have already been written elsewhere in Tincr.
- Tincr has unit tests! They are built on the [tcltest](http://www.tcl.tk/man/tcl/TclCmd/tcltest.htm) framework. I began documenting it on [the wiki](https://github.com/byuccl/tincr/wiki/tincr_test), but didn't get very far unfortunately. There should be a couple tests in there already that you can use as examples.

ttown523 committed:
Merge branch 'master' of https://github.com/byuccl/tincr

ttown523 commented:
Brad,

Thank you for taking the time to review this pull request. I have added some comments directly to your feedback, but I am going to add more general thoughts on your feedback here:

**Caching**: I agree with you that caching would be an improvement to the script, but for us right now it is low priority for these reasons:

1. The performance of the script is not bad enough to require caching to make it usable. These cell libraries need to only be generated once, and run on the order of minutes (the part I have been working with generates in ~1 minute). 
2. As I commented in issue #33 there are still some caches that are not working correctly. 
3. We would like to do an alpha release of RapidSmith 2.0 soon, and we want to provide the users with a stable genCellLibrary.tcl script as soon as possible. 

For these reasons, I think that we should push back the caching to a later date (add an issue to update things with caches)? Again, I agree that it would improve the quality and performance of the script but it is not immediately necessary. 

**Moving methods**: I moved several methods to their appropriate ensembles.

**Write xml methods**: The format of the XML output files is specific to the RapidSmith cell library. It is not for general use, so I think we should keep it in the genCellLibrary.tcl file. I see now use case where a user would want to print out only a portion of a cell library. If they want, they can use the cell library for their own purposes, but in that case they need to print the entire thing out. So, I propose we leave these methods in this file.

**Testing**: Yes, unit tests are a good idea, but after looking at a few of your tests, I need to take the time to learn how to create a unit test. I think that we can create another pull request that contains the unit tests in the future.

I have pushed my new changes so you can review them. 

Thanks again for reviewing this!

ttown523 committed:
suggested changes to pull request #38
